overlord/ifacestate/handlers.go

```diff
@@ -1123,18 +1166,44 @@ func (m *InterfaceManager) doAutoConnect(task *state.Task, _ *tomb.Tomb) error {
 		}
 	}
 
-	autots, err := batchConnectTasks(st, snapsup, newconns, connOpts)
+	autots, hasHooks, err := batchConnectTasks(st, snapsup, newconns, connOpts)
 	if err != nil {
 		return err
 	}
 
-	if m.preseed && len(autots.Tasks()) > 2 { // connect task and setup-profiles tasks are 2 tasks, other tasks are hooks
-		// TODO: in preseed mode make interface hooks wait for mark-preseeded task.
-		for _, t := range autots.Tasks() {
-			if t.Kind() == "run-hook" {
-				return fmt.Errorf("interface hooks are not yet supported in preseed mode")
+	// If hooks are not present then connects can be executed during preseeding.
+	// Otherwise we will run all connects, their hooks and setup-profiles after
+	// preseeding (on first boot). Note, we may be facing multiple connections
+	// here where only some have hooks; however there is no point in running
+	// those without hooks before mark-preseeded, because only setup-profiles is
+	// performance-critical and it still needs to run after those with hooks.
+	if m.preseed && hasHooks {
```

Comment:
just so that I can clarify the ordering here, we have a `mark-preseeded` task, and either an `install` hook or `start-snap-services` task which needs to run after the interface hooks, so what we do is we add all the connect tasks (and also the interface hook tasks) to depend on the `mark-preseeded` task (so it runs after rebooting into seeding), then make the `install` hook or `start-services-task` also depend on all of the auto-connect tasks. Then the ordering is something like:

```
mark-preseeded depends on auto-connect-tasks
install-hook depends on auto-connect-tasks
start-snap-services depends on auto-connect-tasks
```
such that the only possible correct ordering is to run the auto-connect-tasks first, then the install-hook or start-snap-services. Is that an accurate understanding?

Reply:
Yes, your understanding is correct, except for start-services task which is not relevant because we always have install hook task. See my other comments.

Reply:
right I think that was the key thing that I was missing that we _always_ have an install hook task, even if we don't actually have an install hook to actually run.